test/cpp/c10d/ProcessGroupNCCLErrorsTest.cpp

```diff
@@ -141,6 +142,44 @@ class ProcessGroupNCCLTimedOutErrors : public ProcessGroupNCCLSimulateErrors {
   bool set_timedout_error_;
 };
 
+class ProcessGroupNCCLNoHeartbeat : public c10d::ProcessGroupNCCL {
+ public:
+  ProcessGroupNCCLNoHeartbeat(
+      const c10::intrusive_ptr<c10d::Store>& store,
+      int rank,
+      int size,
+      c10::intrusive_ptr<c10d::ProcessGroupNCCL::Options> opts)
+      : ProcessGroupNCCL(store, rank, size, opts) {
+    hasMonitorThreadCaughtError_.store(false);
+  }
+
+  std::mutex& getWatchdogMutex() {
+    return workMetaListMutex_;
+  }
+
+  bool getErrorCaughtFlag() {
+    return hasMonitorThreadCaughtError_.load();
+  }
+
+ protected:
+  void heartbeatMonitor() override {
+    try {
+      c10d::ProcessGroupNCCL::heartbeatMonitor();
+    } catch (std::runtime_error& e) {
+      hasMonitorThreadCaughtError_.store(true);
+    }
+  }
+
+  // It's really hard to unit test std::abort. So we override it instead.
+  // Commented this override, we do see process aborted with core dump.
+  void terminateProcess(std::string errMsg) override {
+    monitorException_ = std::make_exception_ptr(std::runtime_error(errMsg));
```

Comment:
why do you do this 'make_exception_ptr' thing and call rethrow?  can't you just `throw runtime_error(errMsg)`?